src/librustdoc/html/highlight.rs

```diff
@@ -134,151 +135,321 @@ fn can_merge(class1: Option<Class>, class2: Option<Class>, text: &str) -> bool {
+        needs_escape: bool,
+    ) {
+        // If the new token cannot be merged with the currently open `Class`, we close the `Class`
+        // if possible.
+        if !can_merge(self.last_class(), class, &text) {
+            self.close_last_if_needed(out)
+        }
+
+        let current_class = self.last_class();
+
+        // If we have a `Class` that hasn't been "open" yet (ie, we received only an `EnterSpan`
+        // event), we need to open the `Class` before going any further so the new token will be
+        // written inside it.
+        if class.is_none() && !self.last_class_is_open() {
+            if let Some(current_class_info) = self.open_classes.last_mut() {
+                let class_s = current_class_info.class.as_html();
+                if !class_s.is_empty() {
+                    write!(out, "<span class=\"{class_s}\">").unwrap();
+                }
+                current_class_info.open = true;
+            }
+        }
+
+        let current_class_is_open = self.open_classes.last().is_some_and(|c| c.open);
+        let can_merge = can_merge(class, current_class, &text);
+        let should_open_tag = !current_class_is_open || !can_merge;
+
+        let text =
+            if needs_escape { Either::Left(&EscapeBodyText(&text)) } else { Either::Right(text) };
+
+        let closing_tag =
+            string_without_closing_tag(out, &text, class, href_context, should_open_tag);
+        if class.is_some() && should_open_tag && closing_tag.is_none() {
+            panic!(
+                "called `string_without_closing_tag` with a class but no closing tag was returned"
+            );
+        } else if let Some(closing_tag) = closing_tag
+            && !closing_tag.is_empty()
+        {
+            // If this is a link, we need to close it right away and not open a new `Class`,
+            // otherwise extra content would go into the `<a>` HTML tag.
+            if closing_tag == "</a>" {
+                out.write_str(closing_tag).unwrap();
+            // If the current `Class` is not compatible with this one, we create a new `Class`.
+            } else if let Some(class) = class
+                && !can_merge
+            {
+                self.enter_elem(out, href_context, class, true);
+            // Otherwise, we consider the actual `Class` to have been open.
+            } else if let Some(current_class_info) = self.open_classes.last_mut() {
+                current_class_info.open = true;
+            }
+        }
+    }
+
+    fn empty_stack<W: Write>(&mut self, out: &mut W) -> Vec<Class> {
+        let mut classes = Vec::with_capacity(self.open_classes.len());
+
+        // We close all open tags and only keep the ones that were not already waiting to be closed.
+        while let Some(class_info) = self.open_classes.pop() {
+            class_info.close_tag(out);
+            if !class_info.pending_exit {
+                classes.push(class_info.class);
+            }
+        }
+        classes
```

Comment:
```suggestion
        // We close all open tags and only keep the ones that were not already waiting to be closed.
        mem::take(&mut self.open_classes)
            .into_iter()
            .rev()
            .filter_map(|class_info| {
                class_info.close_tag(out);
                (!class_info.pending_exit).then_some(class_info.class)
            })
            .collect()
```

I know you're not a huge iterator fan, so take or leave this suggestion. I *do* think it would be nice to be more explicit about the fact you're emptying `self.open_classes` (took me more than one look to realize that)

Reply:
Hum, I guess it's personal preference. Again I find your code much harder to follow because we have two indirections: `mem::take` and `.rev()`, making it tricky to understand what's actually going on.